### PR TITLE
matcher: Generic matching API: configuration validation

### PR DESCRIPTION
**Pull request by tyxia**
With the generic matching API introduced in PR #26861, data input type and input matcher type can be types other than string. (It is only string for now) 

**Adding the configuration validation:**
Detect the mismatch between input type and supported input type of matcher. For example, technically `string` matcher should not have `float` type input. 

1. Add `supportedDataInputTypes` to `InputMatcher`: Even though currently input and matcher is 1:1, forward thinking matcher could support multiple input types. So it returns a set of supported types.
2. Add `dataInputType` to `DataInput`: One `DataInput` should have one input type. So returns typeid of supported types.

**For list matcher:**  
There are 4 [matcher types](https://github.com/envoyproxy/envoy/blob/9b7144374cca05b029fdcac816bf81eb87e6f54e/api/envoy/config/common/matcher/v3/matcher.proto#L80-L89): single_predicate, or_matcher (OR multi_predicates), and_matcher (AND multi_predicates), not_matcher. They will eventually call `SingleFieldMatcher` to CreateFieldMatcher. So adding check in the constructor of `SingleFieldMatcher` which is centralized place. They all only support string type currently.

CEL matcher will use [custom_match](https://github.com/envoyproxy/envoy/blob/9b7144374cca05b029fdcac816bf81eb87e6f54e/api/envoy/config/common/matcher/v3/matcher.proto#L67) and will introduce its own rule by overriding those methods above when it is added.

**For tree matcher:**  
[map_matcher](https://github.com/envoyproxy/envoy/blob/9b7144374cca05b029fdcac816bf81eb87e6f54e/api/envoy/config/common/matcher/v3/matcher.proto#L121-L124) (exact_match and prefix_match): the input type is fixed to be string because string is used as key in [MatchMap](https://github.com/envoyproxy/envoy/blob/9b7144374cca05b029fdcac816bf81eb87e6f54e/api/envoy/config/common/matcher/v3/matcher.proto#L108). And there is no need for other types currently. Thus, add the check in its matcher to see if input type is string. 

[custom_match](https://github.com/envoyproxy/envoy/blob/9b7144374cca05b029fdcac816bf81eb87e6f54e/api/envoy/config/common/matcher/v3/matcher.proto#L67): It doesn't have fixed rule and it is up to matcher implementation. Currently, the only implemented custom matcher is trie matcher. Even though its input proto type can be `stringVaule` or `v3.DestinationIPInput`, its data input type in c++ is only string.

**Tests:**
Provide the negative test cases float input + string matcher hit all the exceptions that are added in this PR.
Cover both tree matcher : exact + prefix _ trie and list matcher : single predicate + multi-predicate (and_matcher)

Fixes issue #27007

Signed-off-by: tyxia [tyxia@google.com](mailto:tyxia@google.com)

### CODE REVIEW

File: source/common/matcher/field_matcher.h [hunk 1]
```diff
@@ -167,6 +168,15 @@ class SingleFieldMatcher : public FieldMatcher<DataType>, Logger::Loggable<Logge
   }
 
 private:
+  static DataInputPtr<DataType>& validateInputAndMatcher(DataInputPtr<DataType>& data_input,
+                                                         InputMatcherPtr& input_matcher) {
+    auto supported_dataInput = input_matcher->supportedDataInputTypes();
```
Comment (htuch) [inline source/common/matcher/field_matcher.h:173]: Nit: weird mix of snake/camel casing.

File: envoy/matcher/matcher.h [hunk 2]
```diff
@@ -161,10 +162,19 @@ class InputMatcher {
...
-   * @param absl::optional<absl::string_view> the value to match on. Will be absl::nullopt if the
+   * @param Matcher::MatchingDataType the value to match on. Will be absl::monostate() if the
    * lookup failed.
    */
   virtual bool match(const Matcher::MatchingDataType& input) PURE;
+
+  /**
+   * A set of data input types supported by InputMatcher.
+   * String is default supported data input type because almost all the derived objects support
+   * string only. Override this function to provide matcher specific supported data input types.
```
Comment (htuch) [inline envoy/matcher/matcher.h:173]: Can you provide guidance on what non-default input types should look like? In the default case it's typeid of string. Is it expected to be the typeid for other things too?

**Review by htuch — COMMENTED** [review]
Basic approach makes sense to me, over to @wbpcode for their take.

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (wbpcode) [reply envoy/matcher/matcher.h:173]: Seems they are hard to read for human. But may be we needn't to read the final string?

```
int: i
std::string: NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
A: N12_GLOBAL__N_11AE
B: N2NA1BE
```

Comment (tyxia) [reply envoy/matcher/matcher.h:173]: Re @htuch: yes, the basically it is typeid for type of [data input result](https://github.com/envoyproxy/envoy/blob/fa5f548d05da6495c175b5c124292fa0bae215de/envoy/matcher/matcher.h#L206) and the input to [input matcher](https://github.com/envoyproxy/envoy/blob/fa5f548d05da6495c175b5c124292fa0bae215de/envoy/matcher/matcher.h#L164) (the type of both is same). For example, typeid(int), typeid(float), typeid(CustomObject) like
typeid(StreamActivation). In my next PR for CEL, [StreamActivation](https://github.com/tyxia/envoy/blob/3f6ed4811e9fc6d92198b72bedb7bfd9d6215623/source/extensions/filters/common/expr/evaluator.h#L27) is the CEL context returned from `dataInput` and used in `inputMatcher` to invoke [Evaluate](https://github.com/tyxia/cel-cpp/blob/ea339f1f73818afe5699fbef8796a4c0c617b5f7/eval/public/cel_expression.h#L49)

Re @wbpcode: Yea, the return of typeid().name is not human readable. I don't think we will read this carefully. 
As long as it is consistent way of retrieving type between dataInput and inputMatcher and it is standard output (i.e., can be looked up if people want), it should be fine.

File: source/common/matcher/field_matcher.h [hunk 3]
```diff
@@ -144,7 +144,8 @@ template <class DataType>
 class SingleFieldMatcher : public FieldMatcher<DataType>, Logger::Loggable<Logger::Id::matcher> {
 public:
   SingleFieldMatcher(DataInputPtr<DataType>&& data_input, InputMatcherPtr&& input_matcher)
-      : data_input_(std::move(data_input)), input_matcher_(std::move(input_matcher)) {}
+      : data_input_(std::move(validateInputAndMatcher(data_input, input_matcher))),
+        input_matcher_(std::move(input_matcher)) {}
```
Comment (wbpcode) [inline source/common/matcher/field_matcher.h:148]: > may just validate the input and matcher in the constructor function rather than in the initilize list to improve the readability?

sorry, i think this is comment for this.

File: source/common/matcher/exact_map_matcher.h [hunk 4]
```diff
@@ -13,7 +13,8 @@ template <class DataType> class ExactMapMatcher : public MapMatcher<DataType> {
 public:
   ExactMapMatcher(DataInputPtr<DataType>&& data_input,
                   absl::optional<OnMatch<DataType>> on_no_match)
-      : MapMatcher<DataType>(std::move(data_input), std::move(on_no_match)) {}
+      : MapMatcher<DataType>(std::move(MapMatcher<DataType>::validateDataInput(data_input)),
+                             std::move(on_no_match)) {}
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply source/common/matcher/exact_map_matcher.h:17]: and for the MapMatcher, could we just validate it in the `MapMatcher`'s constructor?

File: source/common/matcher/field_matcher.h (hunk 3 above)
Comment (tyxia) [reply source/common/matcher/field_matcher.h:148]: Yea, here we can do it on `data_input_` and `input_matcher_ `. But i didn't do this because:
1) be consistent with `initializer_list_validate style` at places above where we cannot do validation in constructor.
2) slightly better to throw exception earlier and don't initialized those class member when it is invalid case. i.e., initialization of class members will be redundant for invalid case.

wdyt?

File: source/common/matcher/exact_map_matcher.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (tyxia) [reply source/common/matcher/exact_map_matcher.h:17]: Yea, it is a good idea to do it in MapMatcher. Will update. 
I slightly prefer `initializer_list_validate style` as I explained in comment below though. I think Envoy code base has this style (need to search the codebase to find the example)

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (yanavlasov) [reply envoy/matcher/matcher.h:173]: Technically i think it can be anything. It does not need to be typeid(). typeid() depends on RTTI being enabled, which I think at this point is a requirement for building Envoy.

I would just define a few human readable strings for core supported types, in case anyone needs to debug this in the future.

File: source/common/matcher/field_matcher.h [hunk 5]
```diff
@@ -167,6 +168,15 @@ class SingleFieldMatcher : public FieldMatcher<DataType>, Logger::Loggable<Logge
   }
 
 private:
+  static DataInputPtr<DataType>& validateInputAndMatcher(DataInputPtr<DataType>& data_input,
+                                                         InputMatcherPtr& input_matcher) {
+    auto supported_input_types = input_matcher->supportedDataInputTypes();
+    if (supported_input_types.find(data_input->dataInputType()) == supported_input_types.end()) {
+      throw EnvoyException(
+          absl::StrCat("Unsupported data input type: ", data_input->dataInputType()));
```
Comment (yanavlasov) [inline source/common/matcher/field_matcher.h:176]: Given that this error is user visible, this is makes it more important to use human readable string for type identification. It would also be good to include both the expected and offered types.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/common/matcher/field_matcher.h (hunk 3 above)
Comment (wbpcode) [reply source/common/matcher/field_matcher.h:148]: I think envoy have no any preference to the style. Both styles are ok.

Specific to this patch, imo, validate the input type in the ctor function will make the code easy to read/understand. This is very important for long term maintenance.

By that way, we needn't prapagate the input to the validator and return it after the validation and finally move it.

And about the overhead, I think there should no any difference.

File: source/common/matcher/field_matcher.h (hunk 5 above)
Comment (wbpcode) [reply source/common/matcher/field_matcher.h:176]: Sounds reasonable, esp for testing and debugging.

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (tyxia) [reply envoy/matcher/matcher.h:173]: Yea, for core types like std::string, I can define readable name. Just to confirm for custom object like CEL context, we are still using typeid? If not, then every time a new custom object is added, the name definition need to be added as well, which might not be ideal.

Comment (tyxia) [reply envoy/matcher/matcher.h:173]: To improve the naming readability. ideally, we can use util::Demangle (which seems not be available in OSS Envoy) or pretty print lib like boost/type_index.  In general, I think we should avoid hardcode if possible. But for those a few core types, it should be fine to have literal name for them
Will update.

File: source/common/matcher/field_matcher.h [hunk 6]
```diff
@@ -144,7 +144,19 @@ template <class DataType>
...
   SingleFieldMatcher(DataInputPtr<DataType>&& data_input, InputMatcherPtr&& input_matcher)
-      : data_input_(std::move(data_input)), input_matcher_(std::move(input_matcher)) {}
+      : data_input_(std::move(data_input)), input_matcher_(std::move(input_matcher)) {
+    auto supported_input_types = input_matcher_->supportedDataInputTypes();
+    if (supported_input_types.find(data_input_->dataInputType()) == supported_input_types.end()) {
+      std::string supported_types;
+      for (const auto& str : supported_input_types) {
+        absl::StrAppend(&supported_types, str, ", ");
+      }
+      supported_types.erase(supported_types.length() - 2);
```
Comment (wbpcode) [inline source/common/matcher/field_matcher.h:154]: `absl::StrJoin`?

File: source/common/matcher/field_matcher.h (hunk 3 above)
Comment (tyxia) [reply source/common/matcher/field_matcher.h]: Here is 2 std::move on smart pointers. So, the overhead is not just moving two pointers as it involves updating the internal state to transfer ownership.

But I don't have strong preference over `initializer_list_validate` style either.  Either style works for me and

File: source/common/matcher/field_matcher.h (hunk 6 above)
Comment (tyxia) [reply source/common/matcher/field_matcher.h]: good idea. Done

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (htuch) [reply envoy/matcher/matcher.h:173]: @tyxia I think at least update the code comments to convey what is really going on right now.

Comment (tyxia) [reply envoy/matcher/matcher.h]: I have updated the comment for core types. For future custom type like cel matcher, I will try to see if I can avoid hardcoded name. I am thinking of introducing demangle util via abi::__cxa_demangle(which is available in clang).
Thus I will update here again if needed when I introduce the CEL matcher

File: envoy/matcher/matcher.h [hunk 7]
```diff
@@ -25,7 +26,7 @@ class ServerFactoryContext;
 namespace Matcher {
 
 using MatchingDataType = absl::variant<absl::monostate, std::string>;
-
+const char DefaultMatchingDataType[] = "string";
```
Comment (wbpcode) [inline envoy/matcher/matcher.h:29]: If we think this type name should be pre-defined or pre-declared, may be `constexpr absl::string_view` would be better.

And the `supportedDataInputTypes` could return the set of string view. The `dataInputType`  could also return a  string view.

For a big matcher, this way could reduce lots of unnecessary string copy or memeory allocation in the config loading phase. (With the more and more complex configuration, I think sometimes we should also concern the performance at the config/xds loading)

**Review by wbpcode — COMMENTED** [review]
Only final comment. others lgtm.

Although I think we should figure out a better to declare the type name, but we can never complete all thing in one PR.

File: envoy/matcher/matcher.h (hunk 7 above)
Comment (tyxia) [reply envoy/matcher/matcher.h:29]: IMHO, the string_view is one of common over-used concepts, especially for function return type and stored in a container or class member, which introduces unnecessary concern about lifetime but doesn’t save anything. Because 1)in most cases of std::string, there are actually no copies at all and 2)In the case when the copy of std::string is required because the lifetime of object must be guaranteed, string_view doesn't save anything since it requires the copy when it is converted to std::string() 

For the definition here, there is no difference between `const char[]` and `string_view`. string_view is just like char*, a pointer with length.
For copy, like the example below, there is no copy at all c++17 onwards, In earlier language modes, this is usually RVO (go/totw/11) and is guaranteed to be at worst a move. ```
auto input_type = data_input_->dataInputType();
```

In summary, `string_view` is usually suggested for function arguments (probably the only place safe to use without careful consideration.) but not encouraged in containers or class members. Storing string_view is source of programmer error. More details can be found https://abseil.io/tips/180. Also here https://abseil.io/tips/101 about copies.

Comment (yanavlasov) [reply envoy/matcher/matcher.h:29]: absl::string_view provides you with a stronger type safety and optimizations. It is almost always preferable over `char[]` or `char*`. For example, down below you have this code:

```
 auto input_type = data_input_->dataInputType();
 if (input_type != DefaultMatchingDataType) {
```

what would  happen with the existing code like above if the dataInputType() method is changed line this?

```
 virtual const char* dataInputType() const
```

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: envoy/matcher/matcher.h (hunk 7 above)
Comment (tyxia) [reply envoy/matcher/matcher.h:29]: Re @yanavlasov  I don't think people should change `dataInputType` to `const char*` in the first place. IMHO, that is a hypothetical assumption that probably won't and should not happen.  I am neutral on this though as it is just simple const declaration.

Re @wbocope : I don't think we should use `string_view` in the container as it is source of programmer error. In terms of performance, I think benchmark to prove the performance overhead is needed. We should not do micro-optimization the code (just for 6 chars "string") without real benchmark data while introducing the bad programming principle. Also, as i mentioned before we should not hardcode/pre-declared for custom matcher. I think demangle typeid is better approach. So, I am not convinced to change to string_view.

Comment (wbpcode) [reply envoy/matcher/matcher.h]: And you can do a quick search to the code tree, there should lots of method to return string view and also lots of method to return container with string view.

the lifetime problem is important. But in this scenario, for the static string to represent a type, we need worry it.

Comment (tyxia) [reply envoy/matcher/matcher.h]: I don't think those `return string view and also lots of method to return container with string view` means they are good example. In contrast, IMHO I think we should follow general prince https://abseil.io/tips/180 which make more sense.

Comment (wbpcode) [reply envoy/matcher/matcher.h]: Yeah, it is good tips. And should be obeyed always.

But I don't know why return string view violate this rule. esp we know exactly there will no dangling reference.

Basically string view is a reference, and obviously returning references is a very common practice.

The key point is no dangling reference, not no reference.

Comment (tyxia) [reply envoy/matcher/matcher.h]: Return string view doesn't violate this. But from my past experience , many places on the caller site you will still need to convert string_view to string to extend its life time.
 
Returning reference doesn't save things as you may expect but introduce lifetime concern in general. Nowadays compiler is smart to optimize away many copies if you just return by value. https://abseil.io/tips/101  We always need extra care about reference like pointer.

IMHO, robust/easy programming is also encouraged and important in C++. Same philosophy like raw pointer should be more performant than smart pointer but it is not safe to use and not encouraged in modern c++ nowadays.

Comment (wbpcode) [reply envoy/matcher/matcher.h]: So, it depends. In some scenarios string should be returned.

And in this scenario, no lifetime problem, no need extend the lifetime, it's only be used for one time searching after it be returned. This scenario shoud be the case where string view could be used.

Smart pointer is different with string. Smart pointer is cheap, unique pointer is especially cheap. But string is expensive.

Comment (wbpcode) [reply envoy/matcher/matcher.h]: Anyway, seems there's tech disagreement.

But it doesn't matter. I will mark this as unresolved and let other maintainers to make the decision.

Except this point, others lgtm. Thanks for your contributions.

Comment (tyxia) [reply envoy/matcher/matcher.h]: I don't think string is cheaper than smart pointer or cost of them two are comparable. string has small string optimization. There is no doubt that smart pointer is more expensive than raw pointer as it  involves extra overhead (like the cost of creating and managing the unique pointer object itself or ref_count of shared_pontter).

My smart pointer is just an example to prove why the robust and simple programming principle is desired.  People should not spend time to figuring out whether it is safe or not for very specific when there is no benchmark data proves the performance gain or gain is negligible.

Comment (tyxia) [reply envoy/matcher/matcher.h]: As long as we don't store the returned string_view of function, for the case here that is `inline constexpr absl::string_view` (to ensure safe const in header), string_view will force a copy to string on the caller site to ensure the lifetime when needed. So, i am ok with either returning string or string_view in this particular case though i don't think returning string_view/reference is encouraged way in general as described above.

Comment (wbpcode) [reply envoy/matcher/matcher.h]: Typically, if there is no special liftetime problem, the object should return const string& or string view to avoid unnecessary new string creation.
The object should ensure the view/ref has liftetime no shorter than the object self.
**This is common practice rather than specially optimization.**

(RVO is helpful for case where we intend to create new string/object, but not enough if we can complete avoid the creation)

**Review by wbpcode — APPROVED** [review]
LGTM, thanks. Defer this to @yanavlasov for final review/merge. ci is flaky, but should not related to this PR.

File: envoy/matcher/matcher.h (hunk 7 above)
Comment (tyxia) [reply envoy/matcher/matcher.h]: > The object should ensure the view/ref has liftetime no shorter than the object self. This is common practice rather than specially optimization.

I don't think the view/ref's lifetime can be guaranteed in C++. Otherwise, raw pointer should be always used (It should be used when there is no ownership transfer ). So same to raw pointer, every time people see the ref, they will need to think a bit more about the lifetime of the object it refers to. Yes, the static/global objects lifetime is guaranteed but it should be very few of these object in production code. An common use case is to return ref of class member in class method and this is because this lifetime of method is guaranteed to same as class object. Unless the guarantees like example above, I don't think return ref is encouraged.  Very common mistake of new programer is returning the ref of local variable.

I agree the performance optimization is good but making code more complicated when there is no performance gain in first place is not good tradeoff. Not just RVO. For example `Returning std::string, initializing std::string` In C++17 onwards, no copy or move is performed . In earlier language modes, this is usually RVO and is guaranteed to be at worst a move.

Comment (wbpcode) [reply envoy/matcher/matcher.h]: >  Another common use case is to return ref of class member in class method and this is because this lifetime of method is guaranteed to same as class object.

I think I mainly mean this scenario. Return the ref of class member, ref of parent class member, global object, etc.

I agree we shouldn't make the code more complicated for uncertain performance improvement.
But typically we cannot agree on what is complicated code for now.

We do too much disscussions here about this topic, just let's move to the next.

**Review by yanavlasov — APPROVED** [review]
